Clearly separate fixes and CMOR checks

schlunma opened the pull request:
<!--
    Thank you for contributing to our project!

    Please do not delete this text completely, but read the text below and keep
    items that seem relevant. If in doubt, just keep everything and add your
    own text at the top, a reviewer will update the checklist for you.

-->

## Description

### Current situation

At the moment, automatic fixes to datasets are done in the [`CMORCheck` ](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/api/esmvalcore.cmor.html#esmvalcore.cmor.check.CMORCheck) class via the `automatic_fixes` keyword. While this weird interconnection is not only conceptually awkward  (why would we fix and check at the same time?), this also brings along performance issues since CMOR checks are done twice or even more often (https://github.com/ESMValGroup/ESMValCore/issues/2154) for a given dataset. Especially for datasets with lots of individual files, this can add up (e.g., I found a 14% decrease in run time for an ICON recipe).

### Changes in this PR

This PR disentangles automatic fixes and CMOR checks in a fully backwards-compatible way (apart from the deprecations; see below).

All automatic fixes have been removed from the `CMORCheck` class and the `automatic_fix` keyword deprecated (see below). Moreover, a new class [`AutomaticFix`](https://github.com/ESMValGroup/ESMValCore/blob/0d4dccf163cc3568d8f4b34bccd49ed2b1dd6c5d/esmvalcore/cmor/_fixes/automatic_fix.py#L254) has been introduced that now takes care of ALL automatic fixes, which is used in the [`fix_metadata`](https://github.com/ESMValGroup/ESMValCore/blob/0d4dccf163cc3568d8f4b34bccd49ed2b1dd6c5d/esmvalcore/cmor/fix.py#L176-L180) and [`fix_data`](https://github.com/ESMValGroup/ESMValCore/blob/0d4dccf163cc3568d8f4b34bccd49ed2b1dd6c5d/esmvalcore/cmor/fix.py#L295-L299) functions. The CMOR checks within these fix functions are also deprecated (see below).

### Deprecations

- Option `automatic_fixes` of [`CMORCheck`](https://esmvaltool--2157.org.readthedocs.build/projects/ESMValCore/en/2157/api/esmvalcore.cmor.html#esmvalcore.cmor.check.CMORCheck): use the functions [`fix_metadata`](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/api/esmvalcore.preprocessor.html#esmvalcore.preprocessor.fix_metadata), [`fix_data`](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/api/esmvalcore.preprocessor.html#esmvalcore.preprocessor.fix_data), or [`Dataset.load`](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/api/esmvalcore.dataset.html#esmvalcore.dataset.Dataset.load) (which automatically includes the first two functions) instead. CMOR Checks will not include fixes anymore in the future.
- Option `check_level` in [`fix_metadata`](https://esmvaltool--2157.org.readthedocs.build/projects/ESMValCore/en/2157/api/esmvalcore.preprocessor.html#esmvalcore.preprocessor.fix_metadata) and [`fix_data`](https://esmvaltool--2157.org.readthedocs.build/projects/ESMValCore/en/2157/api/esmvalcore.preprocessor.html#esmvalcore.preprocessor.fix_data): use the functions [`cmor_check_metadata`](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/api/esmvalcore.preprocessor.html#esmvalcore.preprocessor.cmor_check_metadata), [`cmor_check_data`](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/api/esmvalcore.preprocessor.html#esmvalcore.preprocessor.cmor_check_data), or [`cmor_check`](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/api/esmvalcore.cmor.html#esmvalcore.cmor.check.cmor_check) instead. Fixes will not check the data anymore in the future.

***

<!--
    Please describe your changes here, especially focusing on why this pull
    request makes ESMValCore better and what problem it solves.

    Before you start, please read our contribution guidelines: https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html

    Please fill in the GitHub issue that is closed by this pull request,
    e.g. Closes #1903
-->

Closes #2154

***

## [Before you get started](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#getting-started)

- [x] [☝ Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do

## [Checklist](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#checklist-for-pull-requests)

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

<!-- The next two lines turn the 🛠 and 🧪 below into hyperlinks -->
[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [x] [🧪][2] The new functionality is [relevant and scientifically sound](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#scientific-relevance)
- [x] [🛠][1] This pull request has a [descriptive title and labels](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-title-and-label)
- [x] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#code-quality)
- [x] [🧪][2] and [🛠][1] [Documentation](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#documentation) is available
- [x] [🛠][1] [Unit tests](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#tests) have been added
- [x] [🛠][1] Changes are [backward compatible](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#backward-compatibility) But see deprecations above.
- [x] [🛠][1] The [list of authors](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#list-of-authors) is up to date
- [x] [🛠][1] ~All [checks below this pull request](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-checks) were successful~ Codacy issues can/will not be fixed (the affected functions already had too much arguments).

***

To help with the number pull requests:

- 🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValCore/pulls) in this repository

schlunma commented:
@ESMValGroup/technical-lead-development-team this is ready for review now.

As already announced in https://github.com/ESMValGroup/ESMValCore/issues/2154 this has become a massive PR now, but I think it's worth it. The separation of fixes and CMOR checks is much clearer now, and after removing the checks from the fixes in v2.12 this will lead to much faster run times for ICON-related recipes.

I tested this extensively with many recipes, but didn't find any issues with it. Apart from the deprecations, all changes here are fully backwards-compatible. The output from running the recipes is exactly identical to the current main branch.

Thanks, and sorry again for this massive PR. 🍻

bouweandela commented on tests/integration/cmor/test_fix.py:
```diff
@@ -0,0 +1,910 @@
+    mock.return_value = MockChecker
+
+
+class TestAutomaticFix:
```
I know that classes can be used for grouping tests in a file, but since all tests in this file are part of this single class I don't really see the added benefit. What is the advantage of using a class here?

schlunma replied:
No particular reason, I think it's just nicer to group all these tests in one class. In addition, having all fixtures available with `self` is also quite nice 👍

bouweandela replied:
I'm not sure if I agree, but it's up to you.

bouweandela commented on tests/integration/cmor/test_fix.py:
```diff
@@ -0,0 +1,910 @@
+        )
+
+        # Create sample data with CMOR errors
+        self.time_coord = DimCoord(
```
```suggestion
        time_coord = DimCoord(
```
It looks like this (and many of the other attributes set in this function) are only used inside this function, so why are they prefixed with `self`?

bouweandela reviewed (changes requested):
I really like how this cleans up our code! I did not look at all the changes in detail but tried to focus my review on the overall structure and design choices.

schlunma commented on esmvalcore/cmor/fix.py:
```diff
@@ -164,16 +189,28 @@ def fix_metadata(cubes,
-                                    fail_on_error=False,
-                                    automatic_fixes=True)
+
+        # Automatic fixes
+        automatic_fixer = AutomaticFix.from_facets(
+            project, mip, short_name, frequency=frequency
+        )
```
Very good point, I haven't thought of this. Will try to change that! But please note that this will introduce a lot more changed files since I need to adapt each `get_fixes` call in the tests (and there are a lot of them!).

schlunma replied:
Done in https://github.com/ESMValGroup/ESMValCore/commit/6954d2168280dd2ccbca0e782815d99041072dd4

schlunma commented on esmvalcore/cmor/check.py:
```diff
@@ -196,39 +186,45 @@ def check_metadata(self, logger=None):
 
         self._check_coords_data()
 
+        self.report_debug_messages()
```
Yes, this is a tiny bugfix - `check_data` did not previously report debug messages. There aren't any at the moment, that's why no one noticed, but we should have this call here in case someone adds a debug message at some point.

schlunma commented on esmvalcore/cmor/table.py:
```diff
@@ -72,18 +72,50 @@ def _get_mips(project: str, short_name: str) -> list[str]:
+            f"tables are available: {', '.join(CMOR_TABLES)}."
+        )
+
+    # CORDEX X-hourly tables define the mip as ending in 'h' instead of 'hr'
+    if project == 'CORDEX' and mip.endswith('hr'):
+        mip = mip.replace('hr', 'h')
```
No, I just moved this code from the `get_cmor_checker` function here:

https://github.com/ESMValGroup/ESMValCore/blob/09e20d41e4590f5b146672727964ab1d0849224b/esmvalcore/cmor/check.py#L1089-L1092

I think it makes much more sense to include this here, so that `get_var_info` always works correctly for CORDEX data.

schlunma commented on esmvalcore/cmor/_fixes/automatic_fix.py:
```diff
@@ -0,0 +1,719 @@
+    return np.array(bounds)
+
+
+class AutomaticFix:
```
Since these fixes are applied to all datasets, how about `AllDatasetsFix`?

bouweandela replied:
That would definitely be more clear! Or how about `GenericFix`?

bouweandela replied:
Awesome!

schlunma commented on esmvalcore/cmor/fix.py:
```diff
@@ -89,54 +92,73 @@ def fix_file(
-                 **extra_facets):
-    """Fix cube metadata if fixes are required and check it anyway.
+def fix_metadata(
+    cubes: CubeList,
```
Fixed in https://github.com/ESMValGroup/ESMValCore/pull/2157/commits/716bf07e18db28e079127c67196af64e1e5982ad

schlunma commented on esmvalcore/cmor/fix.py:
```diff
@@ -89,54 +92,73 @@ def fix_file(
+    check_level: CheckLevels = CheckLevels.DEFAULT,
+    session: Optional[Session] = None,
+    **extra_facets,
+) -> CubeList:
```
Fixed in https://github.com/ESMValGroup/ESMValCore/pull/2157/commits/716bf07e18db28e079127c67196af64e1e5982ad

schlunma commented on esmvalcore/cmor/__init__.py:
```diff
@@ -1 +1,180 @@
 """CMOR module."""
+from __future__ import annotations
```
Done in https://github.com/ESMValGroup/ESMValCore/commit/64c51aff98fc5b55e9a2e220df99be0d9bddfe47

bouweandela replied:
Thanks!

schlunma commented on esmvalcore/cmor/check.py:
```diff
@@ -139,47 +115,61 @@ def __init__(self,
+        # TODO: remove in v2.12
+        self._automatic_fix = AutomaticFix(var_info, frequency=frequency)
+
+    def _uses_unstructured_grid(self):
```
Thanks, I didn't know about that one!

Done in https://github.com/ESMValGroup/ESMValCore/commit/47308754d56abfe299487702d87bd269d14191b9

schlunma commented:
Thanks @bouweandela for the review! :rocket: Apart from renaming `AutomaticFix` I think I addressed all your comments. Please let me know what you think.

bouweandela commented on tests/integration/cmor/_fixes/cmip5/test_bcc_csm1_1_m.py:
```diff
@@ -1,32 +1,28 @@
 def test_get_cl_fix():
     """Test getting of fix."""
     fix = Fix.get_fixes('CMIP5', 'bcc-csm1-1-m', 'Amon', 'cl')
-    assert fix == [Cl(None)]
+    assert fix == [Cl(None), AutomaticFix(None)]
```
Maybe you could write this type of test like this
```python
    assert Cl(None) in Fix.get_fixes('CMIP5', 'bcc-csm1-1-m', 'Amon', 'cl')
```
to avoid having to import AutomaticFix everywhere (and us having to change all of these tests again next time there is an update. If there are a few tests that check the complete list of tests that is returned, that should be enough, right?

schlunma replied:
I actually like that we test that ALL fixes are applied to each model. We have similiar tests for datasets that have an `AllVars` fix. In addition, I spent at least an hour to make these changes yesterday, an I really don't want to go trough 100 files again.

bouweandela commented:
Thanks for making the changes!

bouweandela commented:
> An easy way to fix this is to build the _get_single_cube function into AutomaticFix.fix_metadata.

Sounds fine to me, that is quite similar to the current implementation

bouweandela commented on esmvalcore/cmor/fix.py:
```diff
@@ -163,94 +195,96 @@ def fix_metadata(cubes,
-                                    check_level=check_level,
-                                    fail_on_error=False,
-                                    automatic_fixes=True)
+        dataset_str = f'{project}:{dataset}'
+        cube = _get_single_cube(cube_list, short_name, dataset_str=dataset_str)
```
```suggestion
        cube = cube_list[0]
```
would be cleaner. You're sure that there is only a single cube because GenericFix.fix_metadata ensures that and that is always applied.

schlunma replied:
Done in https://github.com/ESMValGroup/ESMValCore/pull/2157/commits/c5054d750a52c9951908beb64ead81128488dbbf

bouweandela commented:
Thanks @schlunma, LGTM!

valeriupredoi commented:
Superb work, gents - I can merge now if you guys ready-ready 🍺

valeriupredoi reviewed (approved):
Lovely work 👍
